src/lib.rs

```diff
@@ -124,6 +246,317 @@ pub enum DumpError<T> {
+            return Err(DumpError::BadHeaderRead(address, e));
+        }
+
+        let header = match DumpAreaHeader::read_from(&hbuf[..]) {
+            Some(header) => header,
+            None => {
+                return Err(DumpError::BadDumpHeader(address));
+            }
+        };
+
+        if header.magic != DUMP_MAGIC {
+            return Err(DumpError::BadMagic(address));
+        }
+
+        if header.address != address {
+            return Err(DumpError::CorruptHeaderAddress(address));
+        }
+
+        if index == i {
+            return Ok(DumpArea {
+                address: address,
+                length: header.length,
+                agent: DumpAgent::from(header.agent),
+            });
+        }
+
+        if header.next == 0 {
+            return Err(DumpError::InvalidIndex);
+        }
+
+        address = header.next;
+        i += 1;
+    }
+}
+
+//
+// Called by the dump agent proxy to claim a dump area on behalf of a
+// specified agent.  This will look for an area that does not have its agent
+// set to DUMP_AGENT_NONE.  If `claimall` is set, all areas will be claimed
+// (or none).  Areas are always claimed from the first area on.
+//
+pub fn claim_dump_area<T>(
+    base: u32,
+    agent: DumpAgent,
+    mut read: impl FnMut(u32, &mut [u8], bool) -> Result<(), T>,
+    mut write: impl FnMut(u32, &[u8]) -> Result<(), T>,
+) -> Result<Option<DumpArea>, DumpError<T>> {
+    let mut address = base;
+    let mut rval = None;
+
+    let claimall = match agent {
+        DumpAgent::Jefe => false,
+        DumpAgent::Task => true,
+        _ => {
+            return Err(DumpError::InvalidAgent);
+        }
+    };
+
+    let agent: u8 = agent.into();
+
+    const HEADER_SIZE: usize = core::mem::size_of::<DumpAreaHeader>();
+
+    loop {
+        let mut hbuf = [0u8; HEADER_SIZE];
+
+        if let Err(e) = read(address, &mut hbuf[..], true) {
+            return Err(DumpError::BadHeaderRead(address, e));
+        }
+
+        let mut header = match DumpAreaHeader::read_from(&hbuf[..]) {
+            Some(header) => header,
+            None => {
+                return Err(DumpError::BadDumpHeader(address));
+            }
+        };
+
+        if header.magic != DUMP_MAGIC {
+            return Err(DumpError::BadMagic(address));
+        }
+
+        if header.address != address {
+            return Err(DumpError::CorruptHeaderAddress(address));
+        }
+
+        if header.agent == DUMP_AGENT_NONE {
+            //
+            // We have a winner!  Set the agent, and write it back.
+            //
+            header.agent = agent;
+
+            if let Err(e) = write(address, header.as_bytes()) {
+                return Err(DumpError::BadAgentWrite(address, e));
+            }
+
+            //
+            // If we want to claim all areas, keep going -- but keep track
+            // of this first area as it will be the one we return.
+            //
+            if !claimall || address == base {
```

Comment:
Splitting this into `claimall` versus `!claimall` functions may be more clear.  Right now, they're intertwined in a non-obvious way, since their behavior is quite different:

- `claimall` requires starting at the `base` address
- `!claimall` iterates until finding an empty area
- `claimall` iterates over all areas
- `!claimall` finds one area then exits

Reply:
I cleaned this up a bit but didn't go so far as to split it -- will look at this more.

Reply:
This -- coupled with your earlier refactoring suggestion -- really cleaned things up. Thank you!